dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/AbstractEventService.java

```diff
@@ -1409,8 +1408,16 @@ private ProgramStageInstance createProgramStageInstance( ProgramStage programSta
         ImportOptions importOptions )
     {
         ProgramStageInstance programStageInstance = new ProgramStageInstance();
-        programStageInstance.setUid( CodeGenerator.isValidCode( programStageInstanceUid ) ? programStageInstanceUid
-            : CodeGenerator.generateCode() );
+        if ( importOptions.getIdSchemes().getProgramStageInstanceIdScheme().equals( IdScheme.UID ))
+        {
+            programStageInstance.setUid( CodeGenerator.isValidCode( programStageInstanceUid ) ? programStageInstanceUid
+                : CodeGenerator.generateCode() );
+        }
+        else if ( importOptions.getIdSchemes().getProgramStageInstanceIdScheme().equals( IdScheme.CODE ))
+        {
+            programStageInstance.setUid( CodeGenerator.generateCode() );
+            programStageInstance.setCode( programStageInstanceUid );
```

Comment:
It's a bit confusing that you keep calling it `programStageInstanceUid` when it actually is a code in this case.. maybe you can call it just `identifier` or something else?